v3/MmlTree/ts/Attributes.ts

```diff
@@ -0,0 +1,191 @@
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview Implements Attribute class for MmlNodes
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+import {PropertyList, Property} from './Node.js';
+
+/*
+ * A constant for when a property shoudl be inherited from the global defaults lists
+ */
+export const INHERIT = '_inherit_';
+
+/******************************************************************/
+/*
+ * Implements the Attributes class for MmlNodes
+ *  (These can be set explicitly, inherited from parent nodes,
+ *   taken from a default list of values, or taken from global
+ *   defaults.)
+ */
+
+export class Attributes {
+    protected attributes: PropertyList;
+    protected inherited: PropertyList;
+    protected defaults: PropertyList;
+    protected global: PropertyList;
+
+    /*
+     * @param {PropertyList} defaults  The defaults for this node type
+     * @param {PropertyList} global  The global properties (from the math node)
+     *
+     * @constructor
+     */
+    constructor(defaults: PropertyList, global: PropertyList) {
+        this.defaults = defaults;
+        this.inherited = Object.create(defaults);
+        this.attributes = Object.create(this.inherited);
+        this.global = global;
+    }
+
+    /*
+     * @param {string} name  The name of the attribute to set
+     * @param {Property} value  The value to give the named attribute
+     */
+    public set(name: string, value: Property) {
+        this.attributes[name] = value;
+    }
+
+    /*
+     * @param {PropertyList} list  An object containing the properties to set
+     */
+    public setList(list: PropertyList) {
+        Object.assign(this.attributes, list);
+    }
+
+    /*
+     * @param {string} name  The name of the attribute whose value is to be returned
+     * @return {Property}  The value of the named attribute (including inheritance and defaults)
+     */
+    public get(name: string) {
+        let value = this.attributes[name];
+        if (value === INHERIT) {
+            value = this.global[name];
+        }
+        return value;
+    }
+
+    /*
+     * @param {string} name  The value of the attribute whose value is to be returned
+     * @return {Property}  The attribute whse name was given if it is explicit on the
+     *                      node (not inherited or defaulted), null otherwise
+     */
+    public getExplicit(name: string) {
+        if (!this.attributes.hasOwnProperty(name)) {
```

Comment:
Well, it's really an enhanced PropertyList -- one with a prototype chain that inherits from the `inherited` and `default` PropertyLists.  It acts like a PropertyList, but the properties may come from any of three places:  the explicit properties in `attributes`, or the ones in `inherited` or the ones in `defaults`.  This is why we don't have to implement the old `Get` function ourselves; object inheritance does that for us.

So to determine if an attributes has been given explicitly (rather than being inherited), we need to use `hasOwnProperty()`.